input_output.ipynb

```diff
@@ -0,0 +1,919 @@
+    "\n",
+    "\n",
+    "If you really have to use it, do not do so in the middle of long-running computation or your model code but ask for all user inputs upfront. If you need to write an interactive applications however, there are other solutions that are more robust.\n",
+    "\n",
+    "</div>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## File I/O\n",
+    "A second very common I/O operation in most programming languages is reading and writing from files. This is actually more complex than just writing on the console and consists of several steps:\n",
+    "\n",
+    "1. You need to find *where* in the operating system the file is located. This gives you the so-called *path* to the file\n",
+    "2. You need to *open* the file for reading or writing. In some operationg systems, this *locks* the file, so that other users or processes cannot write simultaneously with you. This operation gives you a *file handle* that you can use to read or write.\n",
+    "3. Now , you can read (or write) the contents of the file \n",
+    "4. Finally, you need to *close* the file to make it accessible to other processes again and to finalise the writing: in some implementations, `write` only writes the text to a temporary location in memory (a so-called *buffer*) and only writes the content to the file when you close it. This is done to increase response times, as writing to memory is faster than writing to a file on the disk.\n",
+    "\n",
+    "\n",
+    "As a trick to remember closing files: *a file handle is like a door handle*. \n",
+    "\n",
+    "### Paths\n",
+    "To handle the *where* a file is located, python offers the [pathlib] module which is used to represent *paths* to files in a portable manner. This means that you can use the one python object to represent the location of a file  regardless of the oeprating system and python will convert it into the representation specific to the operating system where you are executing the program. For example, windows uses `\\` as a separator for path elements, while unix/linux use `/`. If you use pathlib, you can write code using only `/` and `pathlib`  will automatically select the correct path separator whe the program runs.\n",
+    "\n",
+    "Let's see how to generate a `path` object pointing to the text file [`hello.txt`](./tutorial/tests/data/hello.txt) in the main directory of this repository. To do so, we use the [`Path`](https://docs.python.org/3/library/pathlib.html#basic-use) class of `pathlib`, whose constructor takes the location of the file as a string\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pathlib as pl\n",
+    "\n",
+    "path = pl.Path(\"./tutorial/tests/data/hello.txt\")\n"
```

Comment:
Please add here that the following also works:

```python
pl.Path(".") / "tutorial/tests/data/hello.txt"
```

Also, this which I find super cool:

```python
pl.Path(".") / "tutorial" / "tests" / "data" / "hello.txt"
```